editoast/src/views/authz.rs

```diff
@@ -83,161 +179,579 @@ async fn whoami(Extension(auth): AuthenticationExt) -> Result<Json<WhoamiRespons
+        .into());
+    }
+    if body.grant.is_some() && body.revoke.is_some() {
+        return Err(AuthzError::BadGrantsOperation {
+            message: "Grant and Revoke cannot be defined at the same time".to_string(),
+        }
+        .into());
     }
 
-    check_user_exists(user_id, &regulator).await?;
+    if let Some(grants) = body.grant {
+        for grant in grants {
+            if grant.resource_type != Resource::Infra {
+                return Err(ResourceError::ResourceTypeNotFound {
+                    resource_type: grant.resource_type.to_string(),
+                })?;
+            }
+            match grant.grant {
+                InfraGrant::Reader => {
+                    authorizer
+                        .grant_infra_reader(grant.subject_id, grant.resource_id)
+                        .await
+                        .map_err(AuthzError::from)?;
+                }
+                InfraGrant::Writer => {
+                    authorizer
+                        .grant_infra_writer(grant.subject_id, grant.resource_id)
+                        .await
+                        .map_err(AuthzError::from)?;
+                }
+                InfraGrant::Owner => {
+                    authorizer
+                        .grant_infra_owner(grant.subject_id, grant.resource_id)
+                        .await
+                        .map_err(AuthzError::from)?;
+                }
+            }
+        }
+    }
+    if let Some(revoke) = body.revoke {
+        for grant in revoke {
+            if grant.resource_type != Resource::Infra {
+                return Err(ResourceError::ResourceTypeNotFound {
+                    resource_type: grant.resource_type.to_string(),
+                })?;
+            }
+            authorizer
+                .revoke_infra_reader(grant.subject_id, grant.resource_id)
+                .await
+                .map_err(AuthzError::from)?;
+
+            authorizer
+                .revoke_infra_writer(grant.subject_id, grant.resource_id)
+                .await
+                .map_err(AuthzError::from)?;
+
+            authorizer
+                .revoke_infra_owner(grant.subject_id, grant.resource_id)
+                .await
+                .map_err(AuthzError::from)?;
+        }
+    }
 
-    regulator
-        .revoke_user_roles(user_id, HashSet::from_iter(roles))
-        .await
-        .map_err(AuthzError::from)?;
-    Ok(axum::http::StatusCode::NO_CONTENT)
+    Ok(StatusCode::NO_CONTENT)
 }
 
 #[cfg(test)]
-mod tests {
+pub mod test {
     use axum::http::StatusCode;
 
-    use crate::views::test_app::TestRequestExt as _;
     use crate::views::test_app::test_app;
 
     use super::*;
+    use crate::models::fixtures::create_small_infra;
+    use crate::views::test_app::TestApp;
+    use crate::views::test_app::TestRequestExt;
+    use editoast_authz::subject::UserInfo;
+    use pretty_assertions::assert_eq;
+    use rstest::rstest;
+    use serde_json::json;
+
+    #[rstest]
```

Comment:
```suggestion
    #[tokio::test(flavor = "multi_thread")] // maybe with worker_threads = 1 as well
```

we're trying to get rid of `rstest` where pssible

Reply:
Don't know why, but when changing to tokio, my tests never ends, even when I removed the `multi_thread`